admin: wcr support for uuid override

nvartolomei opened the pull request:
WCR subsystem uploads cluster manifests periodically to object storage
keyed by cluster UUID and a sequence number.

When a cluster is re-created, it gets a new cluster UUID and the
sequence number is initialized as the highest seen so far in the bucket.

Here is an example of a cluster (assigned <uuid-a> initially) after it
uploaded 3 manifests then was re-created (assigned <uuid-b> this time).

```
/
+- cluster_metadata/
   +- <uuid-a>/manifests/
   |  +- 0/cluster_manifest.json
   |  +- 1/cluster_manifest.json
   |  +- 2/cluster_manifest.json (topic A exists)
   + <uuid-b>/manifests/
      +- 3/cluster_manifest.json (empty)
      +- 4/cluster_manifest.json (empty)
```

Attempting WCR searches for highest sequence number ignoring current
cluster uuid (i.e. <uuid-b). If we run WCR, we will pick
`/cluster_metadata/<uuid-a>/manifests/2/cluster_manifest.json` as source
of truth.

This eventually would lead to creation of another manfiest:

```
/
+- cluster_metadata/
   # ...
   +- <uuid-b>/manifests/
      # ...
      +- 5/cluster_manifest.json (topic A recovered)
```

This algorithm (A1) is known to reach an undesirable result in the
following scenario: Imagine that we start with the same state as above
prior to restore but we do not run restore but instead re-create the
cluster once more. This can happen either because an operator wanted to
restart recovery process from scratch or because a disaster happened in
short span before recovery had a chance to finish/start. And we end with
this picture:

```
/
+- cluster_metadata/
   +- <uuid-a>/manifests/
   |  +- 0/cluster_manifest.json
   |  +- 1/cluster_manifest.json
   |  +- 2/cluster_manifest.json (topic A exists)
   + <uuid-b>/manifests/
   |  +- 3/cluster_manifest.json (empty)
   |  +- 4/cluster_manifest.json (empty)
   + <uuid-c>/manifests/
      +- 5/cluster_manifest.json (empty)
      +- 6/cluster_manifest.json (empty)
```

If we run the same algorithm (A1) then it will pick
`/cluster_metadata/<uuid-b>/manifests/4/cluster_manifest.json`. However,
we want to recover from <uuid-a> because that's where our data lives.

To support this, we add an optional override to cluster recovery
subsystem where user can explicitly choose from which cluster UUID to
restore.

---

This also works with `cloud_storage_cluster_name` and has guardrails to
prevent restoring from a UUID belonging to a different cluster and has
guardrails to prevent restoring from a UUID belonging to a different
cluster.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

nvartolomei committed:
admin: split cluster recovery

Split cluster recovery related functionality into a separate translation
unit to make it easier to work with it/recompile.

nvartolomei committed:
admin: wcr support for uuid override

nvartolomei committed:
cluster: wcr support for uuid override

WCR subsystem uploads cluster manifests periodically to object storage
keyed by cluster UUID and a sequence number.

When a cluster is re-created, it gets a new cluster UUID and the
sequence number is initialized as the highest seen so far in the bucket.

Here is an example of a cluster (assigned <uuid-a> initially) after it
uploaded 3 manifests then was re-created (assigned <uuid-b> this time).

```
/
+- cluster_metadata/
   +- <uuid-a>/manifests/
   |  +- 0/cluster_manifest.json
   |  +- 1/cluster_manifest.json
   |  +- 2/cluster_manifest.json (topic A exists)
   + <uuid-b>/manifests/
      +- 3/cluster_manifest.json (empty)
      +- 4/cluster_manifest.json (empty)
```

Attempting WCR searches for highest sequence number ignoring current
cluster uuid (i.e. <uuid-b). If we run WCR, we will pick
`/cluster_metadata/<uuid-a>/manifests/2/cluster_manifest.json` as source
of truth.

This eventually would lead to creation of another manfiest:

```
/
+- cluster_metadata/
   # ...
   +- <uuid-b>/manifests/
      # ...
      +- 5/cluster_manifest.json (topic A recovered)
```

This algorithm (A1) is known to reach an undesirable result in the
following scenario: Imagine that we start with the same state as above
prior to restore but we do not run restore but instead re-create the
cluster once more. This can happen either because an operator wanted to
restart recovery process from scratch or because a disaster happened in
short span before recovery had a chance to finish/start. And we end with
this picture:

```
/
+- cluster_metadata/
   +- <uuid-a>/manifests/
   |  +- 0/cluster_manifest.json
   |  +- 1/cluster_manifest.json
   |  +- 2/cluster_manifest.json (topic A exists)
   + <uuid-b>/manifests/
   |  +- 3/cluster_manifest.json (empty)
   |  +- 4/cluster_manifest.json (empty)
   + <uuid-c>/manifests/
      +- 5/cluster_manifest.json (empty)
      +- 6/cluster_manifest.json (empty)
```

If we run the same algorithm (A1) then it will pick
`/cluster_metadata/<uuid-b>/manifests/4/cluster_manifest.json`. However,
we want to recover from <uuid-a> because that's where our data lives.

To support this, we add an optional override to cluster recovery
subsystem where user can explicitly choose from which cluster UUID to
restore.

---

This also works with `cloud_storage_cluster_name` and has guardrails to
prevent restoring from a UUID belonging to a different cluster and has
guardrails to prevent restoring from a UUID belonging to a different
cluster.

andrwng commented on src/v/cluster/cluster_recovery_manager.cc:
```diff
@@ -105,18 +106,96 @@ cluster_recovery_manager::initialize_recovery(
+            if (!uses_cluster_id_res.has_value()) {
+                vlog(
+                  clusterlog.debug,
+                  "Error checking WCR cluster name ownership in bucket {}: {}",
```
nit: these are infrequent but seem useful to determine why WCR failed. Log at warn level? Same elsewhere

andrwng commented on src/v/cluster/cluster_recovery_manager.cc:
```diff
@@ -105,18 +106,96 @@ cluster_recovery_manager::initialize_recovery(
+            co_return cluster_manifest_res.error();
+        }
+        manifest = std::move(cluster_manifest_res.value());
+    } else {
```
nit: this is a pretty big method with some non-trivial number of scopes. IMO it'd be easier to follow and verify correctness if we pulled out a `find_manifest_for_recovery()` that was written with early returns instead of else cases

andrwng reviewed (commented):
Pretty much lgtm

paulohtb6 reviewed (dismissed):
lgtm from docs

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--584d8bd7-0bad-48bc-8323-b42440ce873a-->
<!--TEST RESULTS-->
<!--24c86379-ff1b-46bd-88e2-56c8241f6475-->
<!--TEST RESULTS-->
<!--d1e66e6d-c7ed-4bf1-a398-9e139fe5036a-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73177">build#73177</a></summary>


| test_class                    | test_method                     | test_arguments                                                                                                                             | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   | test_history                                                                                                                                                        |
|:------------------------------|:--------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests | test_replication_basic          | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/73177#019996a8-656c-4d22-ab3d-4180ea319019 | FLAKY         | 8/21     | upstream reliability is '83.38762214983714'. current run reliability is '38.095238095238095'. drift is 45.29238 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic  |
| JavaCompressionTest           | test_upgrade_java_compression   | {"compression_type": "gzip"}                                                                                                               | integration | https://buildkite.com/redpanda/redpanda/builds/73177#019996a8-6571-4eb4-9558-25f498eba1e2 | FLAKY         | 20/21    | upstream reliability is '99.74554707379136'. current run reliability is '95.23809523809523'. drift is 4.50745 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression     |
| JavaCompressionTest           | test_upgrade_java_compression   | {"compression_type": "gzip"}                                                                                                               | integration | https://buildkite.com/redpanda/redpanda/builds/73177#019996ae-bd33-4730-abf6-5b3a17f5de54 | FLAKY         | 20/21    | upstream reliability is '99.51690821256038'. current run reliability is '95.23809523809523'. drift is 4.27881 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression     |
| JavaCompressionTest           | test_upgrade_java_compression   | {"compression_type": "lz4"}                                                                                                                | integration | https://buildkite.com/redpanda/redpanda/builds/73177#019996a8-6572-445f-b54d-2bef5c93eb0e | FLAKY         | 20/21    | upstream reliability is '99.72602739726028'. current run reliability is '95.23809523809523'. drift is 4.48793 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression     |
| JavaCompressionTest           | test_upgrade_java_compression   | {"compression_type": "lz4"}                                                                                                                | integration | https://buildkite.com/redpanda/redpanda/builds/73177#019996ae-bd34-4812-8be2-eb27e4b4a739 | FLAKY         | 20/21    | upstream reliability is '99.48186528497409'. current run reliability is '95.23809523809523'. drift is 4.24377 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression     |
| JavaCompressionTest           | test_upgrade_java_compression   | {"compression_type": "snappy"}                                                                                                             | integration | https://buildkite.com/redpanda/redpanda/builds/73177#019996a8-6573-4c02-88a5-022230230f88 | FLAKY         | 20/21    | upstream reliability is '99.74554707379136'. current run reliability is '95.23809523809523'. drift is 4.50745 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression     |
| JavaCompressionTest           | test_upgrade_java_compression   | {"compression_type": "snappy"}                                                                                                             | integration | https://buildkite.com/redpanda/redpanda/builds/73177#019996ae-bd36-4081-8135-b0c067b4d034 | FLAKY         | 20/21    | upstream reliability is '99.74619289340102'. current run reliability is '95.23809523809523'. drift is 4.5081 and the allowed drift is set to 50. The test should PASS    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression     |
| JavaCompressionTest           | test_upgrade_java_compression   | {"compression_type": "zstd"}                                                                                                               | integration | https://buildkite.com/redpanda/redpanda/builds/73177#019996a8-6569-4ee8-beb5-a6750f7575f9 | FLAKY         | 20/21    | upstream reliability is '99.74093264248705'. current run reliability is '95.23809523809523'. drift is 4.50284 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression     |
| JavaCompressionTest           | test_upgrade_java_compression   | {"compression_type": "zstd"}                                                                                                               | integration | https://buildkite.com/redpanda/redpanda/builds/73177#019996ae-bd29-4358-bf4d-9021c7940ce2 | FLAKY         | 20/21    | upstream reliability is '99.74160206718346'. current run reliability is '95.23809523809523'. drift is 4.50351 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression     |
| DataMigrationsApiTest         | test_higher_level_migration_api | null                                                                                                                                       | integration | https://buildkite.com/redpanda/redpanda/builds/73177#019996a8-656e-4dfc-8996-53f760da9611 | FLAKY         | 20/21    | upstream reliability is '99.05123339658444'. current run reliability is '95.23809523809523'. drift is 3.81314 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_higher_level_migration_api |
| PartitionMovementUpgradeTest  | test_basic_upgrade              | null                                                                                                                                       | integration | https://buildkite.com/redpanda/redpanda/builds/73177#019996a8-656a-46b4-8e0e-8009be2dc79f | FLAKY         | 19/21    | upstream reliability is '99.30555555555556'. current run reliability is '90.47619047619048'. drift is 8.82937 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementUpgradeTest&test_method=test_basic_upgrade       |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73359">build#73359</a></summary>


| test_class          | test_method            | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                     | test_history                                                                                                                                             |
|:--------------------|:-----------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------|
| OffsetRetentionTest | test_offset_expiration | null             | integration | https://buildkite.com/redpanda/redpanda/builds/73359#0199a0b9-0f3b-48c4-8395-37d51b2a762a | FAIL          | 0/1      |                                                                                                                                                            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=OffsetRetentionTest&test_method=test_offset_expiration |
| QuickTerminateTest  | test_terminate         | null             | integration | https://buildkite.com/redpanda/redpanda/builds/73359#0199a0bf-5884-4026-974c-88ccd6de3bf9 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=QuickTerminateTest&test_method=test_terminate          |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73453">build#73453</a></summary>


| test_class                          | test_method                | test_arguments              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                              |
|:------------------------------------|:---------------------------|:----------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| QuickTerminateTest                  | test_terminate             | null                        | integration | https://buildkite.com/redpanda/redpanda/builds/73453#0199a537-2177-48b1-b630-95b4b14e6a14 | FLAKY         | 17/21    | upstream reliability is '100.0'. current run reliability is '80.95238095238095'. drift is 19.04762 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=QuickTerminateTest&test_method=test_terminate                           |
| WriteCachingFailureInjectionE2ETest | test_crash_all             | {"use_transactions": false} | integration | https://buildkite.com/redpanda/redpanda/builds/73453#0199a537-217a-446e-8034-1271b87c3e0b | FLAKY         | 18/21    | upstream reliability is '86.65018541409147'. current run reliability is '85.71428571428571'. drift is 0.9359 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all          |
| WriteCachingFailureInjectionTest    | test_unavoidable_data_loss | null                        | integration | https://buildkite.com/redpanda/redpanda/builds/73453#0199a537-2174-4848-8662-0ca2e0206963 | FLAKY         | 17/21    | upstream reliability is '93.84885764499121'. current run reliability is '80.95238095238095'. drift is 12.89648 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionTest&test_method=test_unavoidable_data_loss |
</details>
<!--/body-->

nvartolomei committed:
cluster: simplify error handling for initialize_recovery

`result<cluster::errc, cloud_metadata::error_outcome>>` as return type
is a lazy hack to propagate errors from 2 different subsystems. Not only
we propagate errors the caller can't handle we also propagate an error
via success path.

Solve this by introducing a new error code type for recovery manager,
convert errors appropriately and simplify error handling.

Error means error, success means success.

andrwng commented on tests/rptest/tests/cluster_recovery_test.py:
```diff
@@ -286,6 +286,91 @@ def test_admin_recovery(self):
+        )
+        assert (
+            excinfo.value.response.json()["message"]
+            == "Cluster is misconfigured for recovery. Check logs for details."
```
Not suggesting an immediate change, but this makes me wonder if we should also be propagating the error (or at least some more specific details) to the admin endpoint response.

nvartolomei replied:
Don't disagree but tried to stick to existing conventions for error code propagation. Don't want to spend time here and now for error propagation. Made sure there are comprehensive logs about recovery process though. Net change is positive imho.

andrwng commented on src/v/cluster/cloud_metadata/manifest_downloads.cc:
```diff
@@ -438,4 +438,31 @@ check_bucket_contains_cluster_names(
+    if (download_result == cloud_storage::download_result::success) {
+        co_return true;
+    } else if (download_result == cloud_storage::download_result::notfound) {
+        co_return false;
```
Maybe in this case we should list the uuids that this cluster name owns, or list the cluster names that own this uuid (I know this is a more expensive search, but still might be useful to tell an engineer what's going on)

nvartolomei replied:
Yeah, this could be better UX but I don't think it is worth spending time implementing it and maintaining such code.

andrwng reviewed (approved):
LGTM, my feedback is not blocking

Also please remember generally to add useful commit messages. They are helpful to get context when learning about why features are built as they are, and are useful when reviewing code to make sure reviewers' and reviewee's expectations are aligned.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/73359-->
<!--1389f332-67ab-4202-b93b-304abb7fc41f-->
<!--title-->
## Retry command for [Build#73359](https://buildkite.com/redpanda/redpanda/builds/73359)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/offset_retention_test.py::OffsetRetentionTest.test_offset_expiration
```
<!--/body-->

nvartolomei commented:
@andrwng no changes to the code but added comprehensive commit description where relevant and PR description.
